src/modulefilter.jl

```diff
@@ -0,0 +1,142 @@
+using Logging
+
+import Base.CoreLogging: _min_enabled_level
+
+export @debug2, @debug3, @debug4, Debug2, Debug3, Debug4, ModuleFilterLogger
+
+const Debug2 = Debug - 1
+const Debug3 = Debug - 2
+const Debug4 = Debug - 3
+
+debug_docs = """
+    @debug2 message [key=value | value ...]
+    @debug3 message [key=value | value ...]
+    @debug4 message [key=value | value ...]
+
+Debug-specific logging macros (similar to `@debug`) for increasing levels of verbosity; i.e.
+`@debug2` is meant for debug logging slightly more verbose than `@debug`, and so on with
+`@debug3` and `@debug4`. See [`LoggingExtras.@setupdebuglogging()`](@ref) for additional
+information about how to conveniently enable these precision debug levels
+for logging in specific modules.
+"""
+
+macro debug2(exs...); :(@logmsg Debug2 $(exs...)); end
+macro debug3(exs...); :(@logmsg Debug3 $(exs...)); end
+macro debug4(exs...); :(@logmsg Debug4 $(exs...)); end
+
+@eval @doc $debug_docs :(@debug2)
+@eval @doc $debug_docs :(@debug3)
+@eval @doc $debug_docs :(@debug4)
+
+"""
+    ModuleFilterLogger(mod, level, logger)
+
+Custom logger for controlling the log level of a specific module. Non-specified modules
+should not be affected using this logger. All logs for the specified module will be filtered
+on `level` by this logger before being passed to the child `logger`. See
+`LoggingExtras.@setupdebuglogging()` for more information on how to allow using this
+logger conveniently.
+"""
+struct ModuleFilterLogger{T <: AbstractLogger} <: AbstractLogger
+    mod::Module
+    level::LogLevel
+    logger::T
+end
+
+Logging.handle_message(x::ModuleFilterLogger, args...; kw...) = 
+    Logging.handle_message(x.logger, args...; kw...)
+
+function Logging.shouldlog(x::ModuleFilterLogger, level, _module, args...; kw...)
+    if x.mod == _module
+        # i.e. this logger will handle ALL log filtering for a specific module
+        return level >= x.level
+    end
+    # why the call to min_enabled_level here for child logger?
+    # we're basically bypassing the core logging min_enabled_level check by
+    # passing the min level of any loggers because we want to do the level check
+    # *here* for the module-specific logger, but then also need to do the level check
+    # for other loggers
+    return level >= Logging.min_enabled_level(x.logger) && Logging.shouldlog(x.logger, level, _module, args...; kw...)
```

Comment:
We have an internal abstraction for this.
`comp_shouldlog`
See how it is used in the EarlyFilteredLogger.
https://github.com/JuliaLogging/LoggingExtras.jl/blob/396afbb317c400e3b04a51f5dfd3e1880e087653/src/earlyfiltered.jl#L30-L33